Update to a more general Device and Stream API. Add support for input streams (E.g. microphone). Add default format methods.

mitchmindtree opened the pull request:
This update prepares for adding input stream support by removing the
`Endpoint` type (which only supports output streams) in favour of a more
general `Device` type which may support any number of input or output
streams. Previously discussed at #117.

The name `Voice` has been replaced with the more ubiquitous name
stream. See #118 for discussion on this.

Also introduces a new `StreamData` type which is now passed to the
`EventLoop::run` callback rather than the `UnknownTypeBuffer`.
`StreamData` allows for passing either `Input` data to be read, or
`Output` data to be written. ~~As of this `PR`, the `Input` variant is just
a unit type and is not yet implemented.~~

All backends have been updated for the API changes, ~~however input
stream functionality remains unimplemented across each backend.
That is, most backend implementations of `default_input_device`,
`supported_input_formats` and `default_input_format` are currently
`unimplemented!();` and panic if called. The main aim was to update
to the new API while retaining existing functionality across all backends.~~

The `beep.rs` and `enumerate.rs` examples and the crate-root
documentation have also been updated for the API changes.

Closes #117.
Closes #118.

cc @tomaka (or anyone following CPAL) this is a pretty big change, it would
be great to get your thoughts/feedback on this! I'm totally open to alternative
approaches, this is my attempt at keeping functionality familiar and similar to
the current API while also allowing for input support. I've began implementing
the input API for the coreaudio backend and will be trying to get the remaining
backends implemented over the next couple days too. I don't mind whether
they are a part of this PR or a follow up PR.

EDIT: The new API is now implemented fully across all backends! Emscripten
still needs updating to handle input streams - I'm hoping someone more
familiar with the emscripten tool-chain can do this in a follow-up PR.

tomaka commented on examples/beep.rs:
```diff
@@ -1,55 +1,60 @@
+        match data {
+            cpal::StreamData::Output { buffer } => {
+                match buffer {
+                    cpal::UnknownTypeBuffer::U16(mut buffer) => {
```
Style: this could be merged with the above `match` (even if that means duplicating the `StreamData::Output` variant).

tomaka commented on src/alsa/enumerate.rs:
```diff
@@ -7,21 +7,21 @@ use std::ffi::CStr;
-/// ALSA implementation for `EndpointsIterator`.
-pub struct EndpointsIterator {
+/// ALSA implementation for `Devices`.
+pub struct Devices {
```
Should be `DevicesIter` IMO.

mitchmindtree replied:
Hmmm I tend to avoid adding the `Iter` suffix unless I'm trying to avoid ambiguity between a collection and its iterator. IMO `Devices` is fine in the same way that `std::str::Chars` is OK.

tomaka commented:
Thanks! The change is positive. We should probably make sure that the various backends have a working implementation before we ship a version on crates.io.

Please also add entries in `CHANGELOG.md`!

mitchmindtree commented:
OK, input streams are now be supported across coreaudio, alsa and wasapi! This includes the `Device::default_{input/output}_format`, `Device::supported_input_formats` and `Endpoint::build_input_stream` methods.

I've fixed the `Device::name` method on windows (it now uses `FriendlyName` rather than the GUID).

I've updated the enumerate example to display both default input/output devices as well as default input/output formats and supported input/output formats for all devices.

I've also added a new `record_wav.rs` example that records three seconds of audio from the system's default input device and writes this audio to a wav at `$CARGO_MANIFEST_DIR/recorded.wav`.

All three examples work on the macos (10.10), linux and (two) windows 10 systems I've been using for testing, but more testing is of welcome of course!

I've also updated the CHANGELOG to account for all changes in this PR.

This PR should now be ready for review!

mitchmindtree commented:
@freesig could you test this branch (run each of the examples) on your mac and the windows machine while you have it? Also maybe let us know what mac version you're running (i've only tested on 10.10).

freesig commented:
@mitchmindtree It ran fine on the mac(10.13.3)
Everything worked fine on the windows except record_wav

> thread 'main' panicked at 'Failed to get default input device', src\libcore\option.rs:874:4
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: process didn't exit successfully: `target\release\examples\record_wav.exe` (exit code: 101)

mitchmindtree commented:
Ahh right, I guess you just don't have any input audio devices on the windows machine then (you can check this with the `enumerate` example - it should show None for default input device)? Do you have a mic or audio interface that you could use to test it with? All good if not, that behaviour is expected for that example if there's no inputs to record with.

freesig commented:
Yep record_wav works if I just plug a lead into the computers input.

mitchmindtree commented:
Nice one, thanks for that!

tomaka commented on src/wasapi/stream.rs:
```diff
@@ -0,0 +1,789 @@
+                format_attempt.Format
+            };
+
+            /*
```
Please remove

mitchmindtree commented:
Good spot! Fixed.
